### PR TITLE
Improve GatherTask Reliability

### LINKED ISSUE #67: [Bug]: Tag gather tasks don't seem to complete
### Bug Description

image

image

### How to Reproduce?

image

image

### Expected Behavior

_No response_

### Version

1.0.5

### Mod Loader Version

0.20.2

### Mod Loader

Quilt

### Code of Conduct

- [X] I have searched the issue tracker and confirmed that the issue has not been reported
- [X] I have checked the FAQ (if one is present) and my issue is not listed
- [X] I have verified that I am using the latest version of the mod
- [X] I have verified that I have installed all the required dependencies for the mod
- [X] I have verified that I do not have Optifine installed. We will close issues if we detect Optifine is in your pack. We cannot fix Optifine issues. Please do not report issues with Optifine present.

### PR DESCRIPTION
**Pull request by sisby-folk**
Closes #67

Tldr - only picking up items off the floor worked, they only worked for the full amount, and odysseus was putting out counts as strings, bricking the tasks. we'll PR to odysseus later on if that isn't fixed by the time we're done with heracles.

https://github.com/terrarium-earth/Heracles/assets/55819817/890ba3ef-4f67-4625-b5c0-9ba6dc70f288

picking up, partial picking up, inventory transfers, and `/give` now work.

Inventory transfers were fixed by passing the index of changed slots to an injected inventory interface via a slot mixin, then mixing into setChanged. 

Partial picking up was fixed by just setting the value in storage regardless of whether the target is met.

`/give` was fixed by caching a copy of the stack at the start of `add()`, then using it at return. This means the stack is intact (it's cleared in many cases) and therefore can pass preliminary matching tests for tasks, but the inventory is in the correct post-add state.

### CODE REVIEW

File: common/src/main/java/earth/terrarium/heracles/mixins/common/InventoryMixin.java [hunk 1]
```diff
@@ -11,22 +12,48 @@
...
+    @Unique private void heracles$testProgress(ItemStack stack) {
+        if (this.player instanceof ServerPlayer serverPlayer) {
             QuestsProgress progress = QuestProgressHandler.getProgress(serverPlayer.getServer(), serverPlayer.getUUID());
             progress.testAndProgressTaskType(serverPlayer, Pair.of(Optional.of(stack), serverPlayer.getInventory()), GatherItemTask.TYPE);
         }
     }
+
+    @Inject(method = "add(ILnet/minecraft/world/item/ItemStack;)Z", at = @At("HEAD"))
+    private void heracles$addItemStack(int slot, ItemStack stack, CallbackInfoReturnable<Boolean> cir) {
+        heracles$addedStack = stack.copy();
```
Comment (ThatGravyBoat) [inline common/src/main/java/earth/terrarium/heracles/mixins/common/InventoryMixin.java:41]: Wouldn't this copy mean that consuming actions wouldn't consume?

**Review by ThatGravyBoat — CHANGES_REQUESTED** [review]

File: common/src/main/java/earth/terrarium/heracles/mixins/common/InventoryMixin.java (hunk 1 above)
Comment (sleepingdragoninn) [reply common/src/main/java/earth/terrarium/heracles/mixins/common/InventoryMixin.java:41]: Thankfully not! This should explain it right quick - that stack never even gets passed down past the initial check (and rightfully! we need to check the whole inventory anyway)
image

**Review by ThatGravyBoat — APPROVED** [review]
